### DISCUSSION TITLE
Help with seccomp filter

### DISCUSSION THREAD

**Discussion by AlphaNecron** [criyle/go-judge#62] [Q&A]
I'm trying to make a judge, and seccomp filter is not allowing normal calls though I've added all calls shown in `strace`. Can you help me?

**Comment by criyle** [comment]
Which call are you referring to? Seccomp config is provided by github.com/elastic/go-seccomp-bpf and you may check their example to write a config.

**Reply by AlphaNecron** [reply]
Using your config in cmd/runprog, modified a bit to match C++ calls, and it does not even work.

**Reply by criyle** [reply]
Please add log when running with `./runprog -show-trace-details ./a.out` to allow further investigation.

**Comment by AlphaNecron** [comment]
```tracer started:  41367 <nil>
------  41367  ------
set ptrace option for 41367
ptrace stopped
------  41367  ------
ptrace stopped
------  41367  ------
ptrace seccomp before execve (should be the execve syscall)
------  41367  ------
ptrace stop exec
------  41367  ------
seccomp traced
syscall: 21 access <nil>
check stat:  /etc/ld.so.preload
------  41367  ------
seccomp traced
syscall: 257 openat <nil>
open:  /etc/ld.so.cache r 
------  41367  ------
seccomp traced
syscall: 257 openat <nil>
open:  /usr/lib/libstdc++.so.6 r 
------  41367  ------
seccomp traced
syscall: 257 openat <nil>
open:  /usr/lib/libm.so.6 r 
------  41367  ------
seccomp traced
syscall: 257 openat <nil>
open:  /usr/lib/libgcc_s.so.1 r 
------  41367  ------
seccomp traced
syscall: 257 openat <nil>
open:  /usr/lib/libc.so.6 r 
------  41367  ------
seccomp traced
syscall: 218 set_tid_address <nil>
------  41367  ------
seccomp traced
syscall: 273 set_robust_list <nil>
------  41367  ------
seccomp traced
syscall: 334 rseq <nil>
------  41367  ------
process exited:  41367 0
id:"686492272fd040728de6d36fbbc883ab_hello-world_1682241272_e52284405a631fe8.0"  input:"10 20\n5 9 6 5 8 9 10 6 5 6"  duration:1 Result[2.662ms 4.9 MiB][1.355631ms 3.34794ms]
tracer started:  41384 <nil>
------  41384  ------
set ptrace option for 41384
ptrace stopped
------  41384  ------
ptrace stopped
------  41384  ------
ptrace seccomp before execve (should be the execve syscall)
------  41384  ------
ptrace stop exec
------  41384  ------
seccomp traced
syscall: 21 access <nil>
check stat:  /etc/ld.so.preload
------  41384  ------
seccomp traced
syscall: 257 openat <nil>
open:  /etc/ld.so.cache r 
------  41384  ------
seccomp traced
syscall: 257 openat <nil>
open:  /usr/lib/libstdc++.so.6 r 
------  41384  ------
seccomp traced
syscall: 257 openat <nil>
open:  /usr/lib/libm.so.6 r 
------  41384  ------
seccomp traced
syscall: 257 openat <nil>
open:  /usr/lib/libgcc_s.so.1 r 
------  41384  ------
seccomp traced
syscall: 257 openat <nil>
open:  /usr/lib/libc.so.6 r 
------  41384  ------
seccomp traced
syscall: 218 set_tid_address <nil>
------  41384  ------
seccomp traced
syscall: 273 set_robust_list <nil>
------  41384  ------
seccomp traced
syscall: 334 rseq <nil>
------  41384  ------
process exited:  41384 0
id:"686492272fd040728de6d36fbbc883ab_hello-world_1682241272_e52284405a631fe8.1"  input:"8 22\n6 10 7 6 9 10 7 7"  duration:1 Result[582µs 8.9 MiB][1.163373ms 1.763512ms]


```

**Reply by criyle** [reply]
What if you run `./runprog -show-trace-details -unsafe ./a.out`? It seems like the process quit after the `rseq` sys call

**Reply by AlphaNecron** [reply]
![image](https://github.com/criyle/go-judge/assets/57827456/51926df2-bf13-425f-a5ea-77765811a5e2)

**Reply by criyle** [reply]
You may fork and add `/usr/lib/libstdc++.so.6` into the readable file after [this line](https://github.com/criyle/go-sandbox/blob/master/cmd/runprog/config/config.go#L18), then try again.

**Reply by AlphaNecron** [reply]
![image](https://github.com/criyle/go-judge/assets/57827456/5c181061-df79-4c42-b46f-7790144b8e7f)
Still.

**Reply by criyle** [reply]
You may try to add `-type compiler` to enable extra allowed syscalls defined [here](https://github.com/criyle/go-sandbox/blob/master/cmd/runprog/config/config.go#L173-L186).

**Reply by AlphaNecron** [reply]
Now it works. Thanks.

**Reply by AlphaNecron** [reply]
Is it possible to run without tracing?

**Reply by criyle** [reply]
The main point of tracing is to ensure the program not to access unnecessary files. You may use `-runner ns` to use mount namespace to restrict file access instead. In this mode, program runs in an isolated mount namespace with seccomp filter applied.

**Reply by AlphaNecron** [reply]
Oh okay. Thanks.

**Reply by AlphaNecron** [reply]
![image](https://github.com/criyle/go-judge/assets/57827456/5d46a78f-4907-4379-abc4-93219b5334cb)
Is using ActionTrace as the default action a good idea?

**Reply by AlphaNecron** [reply]
Btw runner.Result.Memory increases after each run, from 4MB to 11MB, is it an intended behavior?

**Reply by criyle** [reply]
If Cgroup v2 enabled or running without root, there's no memory max statistic so `max_rss` is being used. However, this statistic is not clearing when `exec` so it inherits parent process value, which is the memory used by `executor_server`.
